datafusion/optimizer/src/eliminate_cross_join.rs

```diff
@@ -39,102 +41,147 @@ impl EliminateCrossJoin {
+/// * 'select ... from a, b where (a.x = b.y and b.xx = 100) or (a.x = b.y and b.xx = 200);'
+/// * 'select ... from a, b, c where (a.x = b.y and b.xx = 100 and a.z = c.z)
+/// * or (a.x = b.y and b.xx = 200 and a.z=c.z);'
+/// * 'select ... from a, b where a.x > b.y'
+///
 /// For above queries, the join predicate is available in filters and they are moved to
 /// join nodes appropriately
+///
 /// This fix helps to improve the performance of TPCH Q19. issue#78
 impl OptimizerRule for EliminateCrossJoin {
     fn try_optimize(
         &self,
-        plan: &LogicalPlan,
-        config: &dyn OptimizerConfig,
+        _plan: &LogicalPlan,
+        _config: &dyn OptimizerConfig,
     ) -> Result<Option<LogicalPlan>> {
+        internal_err!("Should have called EliminateCrossJoin::rewrite")
+    }
+
+    fn supports_rewrite(&self) -> bool {
+        true
+    }
+
+    fn rewrite(
+        &self,
+        plan: LogicalPlan,
+        config: &dyn OptimizerConfig,
+    ) -> Result<Transformed<LogicalPlan>> {
+        let plan_schema = plan.schema().clone();
         let mut possible_join_keys = JoinKeySet::new();
         let mut all_inputs: Vec<LogicalPlan> = vec![];
-        let parent_predicate = match plan {
-            LogicalPlan::Filter(filter) => {
-                let input = filter.input.as_ref();
-                match input {
-                    LogicalPlan::Join(Join {
-                        join_type: JoinType::Inner,
-                        ..
-                    })
-                    | LogicalPlan::CrossJoin(_) => {
-                        if !try_flatten_join_inputs(
-                            input,
-                            &mut possible_join_keys,
-                            &mut all_inputs,
-                        )? {
-                            return Ok(None);
-                        }
-                        extract_possible_join_keys(
-                            &filter.predicate,
-                            &mut possible_join_keys,
-                        );
-                        Some(&filter.predicate)
-                    }
-                    _ => {
-                        return utils::optimize_children(self, plan, config);
-                    }
-                }
+
+        let can_flatten_inputs = can_flatten_join_inputs(&plan);
+        let parent_predicate = if let LogicalPlan::Filter(filter) = plan {
+            // if input isn't a join that can potentially be rewritten
+            // avoid unwrapping the input
+            let rewriteable = matches!(
+                filter.input.as_ref(),
+                LogicalPlan::Join(Join {
+                    join_type: JoinType::Inner,
+                    ..
+                }) | LogicalPlan::CrossJoin(_)
+            );
+
+            if !rewriteable {
+                return rewrite_children(self, LogicalPlan::Filter(filter), config);
             }
+
+            if !can_flatten_join_inputs(&filter.input) {
```

Comment:
this condition can go earlier to save some cpu?

Reply:
I think the call to `rewrite_children` needs to happen to apply the rule recursively before proceeding

I am not quite sure what else we could save. 

I tried to make this clearer in 030444dc3 with comments and moving another definition of `can_flatten_join_inputs` closer to where it was used.